### DISCUSSION TITLE
start_item() and waiting for a Driver

### DISCUSSION THREAD

**Discussion by jonpovey** [pyuvm/pyuvm#70] [Q&A] [answered]
UVM beginner question: When a sequence calls `start_item()`, is that meant to block until the Driver connected to the Sequencer calls `get_next_item()`? The UVM Cookbook hints at it, but is not explicit, and the timing diagrams in that book are not clear on this. There is mention of "late randomization" and it seems like the point is that `start_item()` should be blocking.

I ask because looking at `uvm_sequencer` in pyuvm, its implementation of `start_item()` puts the item on a queue serviced by `uvm_sequencer.run_phase()` which gets the item and immediately pings its start condition, which will cause `start_item()` to return without waiting for the driver. In fact it doesn't seem like `start_item` + `run_phase` here really do much at all apart from causing the `run_phase` task to get scheduled once before `start_item` returns.

Am I missing something? Maybe this is a temporary early implementation and waiting for the driver is planned for later?

**Comment by raysalemi** [comment]
The return from the blocked `start_item()` means that your sequence_item is the one that will be returned by `get_next_item()`.  The driver has blocked on `get_next_item()` and will remain blocked until you await `finish_item()`.  This gives you a chance to load data into the item using the handle, hence the comment about "late randomization". It really should be 'late stimulus' as it allows you to load up the sequence item based on the state of the TB at that moment.

**Reply by jonpovey** [reply]
What if the sequence calls `start_item()` before the driver calls `get_next_item()`, or if you have multiple sequences all calling `start_item()`? It seems like at the moment all those `start_item()` calls would return immediately (and load up their item with incorrect early state), when they should only return when the driver has called `get_next_item()` and that specific sequence has won arbitration to supply the next item.
I haven't tested this in pyuvm.

**Comment by raysalemi** [comment]
Looking at the code I see your point.  This code preserves the order of requests, but apparently not the timing. I'll go look at the IEEE spec tomorrow and see if I went wrong here. First I'll write a test that fully tests the timing. I suspect you're right that it will fail.

**Comment by raysalemi** [comment]
It would not surprise me to find a bug here. I wrote this from whole cloth because the spec was pretty useless for putting together a working sequencer.

**Comment by raysalemi** [comment, marked as answer]
I'm writing a test for this now and will fix it so the test passes.  Also cleaning up the sequence tests somewhat.

**Comment by raysalemi** [comment]
Thanks for the great catch. 

This test fails:

```Python
        class Seq(uvm_sequence):
            async def body(self):
                op = SeqItem("op")
                DataHolder().start_item_call_time = get_sim_time(units="ns")
                await self.start_item(op)
                DataHolder().start_item_return_time = get_sim_time(units="ns")
                op.data = 0
                await self.finish_item(op)
                DataHolder().datum = (op.result == (op.data + 1))
```

Both variables are set to 4 ns.  This also demonstrates how transactions can now receive a simulation time.
